### PR TITLE
Add timestamps to some Makefile rules

### PR DESCRIPTION
**Pull request by xexyl**
In the Makefiles that do not come from another repo, any rule that had
the lines in the form of:
                     
    ... starting"
                
have been changed to:
            
    ... starting at `date`"

and the same thing for the lines that end with:

    ... ending:"

This does make more noise when things have to be done (but that's why 
not all rules have it) but it can help isolate issues sometimes and it 
matches other rules that already have it.

### CODE REVIEW

**Comment by xexyl** [comment]
Do you want me to do this for the dbg and dyn_array repos? I did it for jparse already. Even so it might be something that can be modified a bit - perhaps some rules should not have it and some should? Perhaps clobber and clean related rules don't need it? Perhaps just test and related rules? I can see that being more useful.

Let me know what you think. I will be doing other things for the rest of the day though.

**Review by lcn2 — APPROVED** [review]
Good stuff, thanks @xexyl

**Comment by xexyl** [comment]
> Goof stuff, thanks @xexyl 

Goof or good? 😀

**Comment by lcn2** [comment]
> Do you want me to do this for the dbg and dyn_array repos? I did it for jparse already. 

Yes, please.  And when you do, add `._.DS_Store` to their `.gitignore` files.

> Even so it might be something that can be modified a bit - perhaps some rules should not have it and some should? Perhaps clobber and clean related rules don't need it? Perhaps just test and related rules? I can see that being more useful.
> 
> Let me know what you think. I will be doing other things for the rest of the day though.

While we do say:

One debugger's noise is another debugger's vital clue.

there is a limit as you imply in your question.

Nevertheless, there are limits.  You will notice that the "leaf Makefiles" in the "makefile tree" from the "other repo" do not print the "starting" / "ending stuff", in part because the higher level `Makefile` indicates the the  the "leaf Makefiles"  is being called (i.e., has "starting" / "ending stuff").

So your "Perhaps clobber and clean related rules don't need it? Perhaps just test and related rules?" is correct for the "leaf Makefiles" and thus applies to the `dbg/Makefile` and `dyn_array/Makefile` that theor clobber and clean related rules don't need it?

Please PR those repos accordingly.

**Comment by xexyl** [comment]
I agree at least as far as I can tell at this time. I will take care of it tomorrow morning. Sleep time for me. 

Good night!

**Comment by xexyl** [comment]
> > Do you want me to do this for the dbg and dyn_array repos? I did it for jparse already.
> 
> Yes, please. And when you do, add `._.DS_Store` to their `.gitignore` files.
> 
> > Even so it might be something that can be modified a bit - perhaps some rules should not have it and some should? Perhaps clobber and clean related rules don't need it? Perhaps just test and related rules? I can see that being more useful.
> > Let me know what you think. I will be doing other things for the rest of the day though.
> 
> While we do say:
> 
> One debugger's noise is another debugger's vital clue.
> 
> there is a limit as you imply in your question.
> 
> Nevertheless, there are limits. You will notice that the "leaf Makefiles" in the "makefile tree" from the "other repo" do not print the "starting" / "ending stuff", in part because the higher level `Makefile` indicates the the the "leaf Makefiles" is being called (i.e., has "starting" / "ending stuff").
> 
> So your "Perhaps clobber and clean related rules don't need it? Perhaps just test and related rules?" is correct for the "leaf Makefiles" and thus applies to the `dbg/Makefile` and `dyn_array/Makefile` that theor clobber and clean related rules don't need it?
> 
> Please PR those repos accordingly.

On the other hand ... those rules already have a `starting` and `ending` message so why not have the timestamps? A question that might be worth asking is if those rules need the starting/ending message at all. In the mkiocccentry repo that would be most useful in the subdirectories, which might mean that the pull requests I made should have them removed (for the clobber/clean rules). But does that mean the message entirely or the date? It seems not a problem to have the date if the message is there. But having in mkiocccentry subdirectories have it (which would happen due to the repos having it) seems a bit much. That's something to consider.

I forgot the .gitignore update .. will do that quickly on the same pull request as I have to do other things. I asked some questions in the issue about getting ready for code freeze but I have other things I must do in the meantime. I might have more time today again but it is not clear how much or what will be done. If the questions are not answered over in that issue then I can't really proceed at least not in the tests part. Maybe documentation can be looked at though.

**Comment by lcn2** [comment]
> On the other hand ... those rules already have a starting and ending message so why not have the timestamps?

A case against adding timestamps to starting and ending messages, is one of more output.  And then, what do you gain from seeing timestamps.  Performance on a given system with a given distribution can be variable.  And even if you do see someone's system, talking longer than expected, what can you do about it?

The tools in this repo are not intended to be "performance zoomers".  The time it takes to do a `make all` is going to be system and computer dependent and will be impacted by whatever else is going on in the system.   However, once `make all` is done once, someone who wishes to submit to the IOCCC does not really need to re-compile.  And the `mkiocccentry(1)` itself is interactive, so performance of that tool is limited by human interaction time: and even then, so what?

Sure someone doing a `make test` might notice how long it takes.  However it someone filed an issue of "`make test` takes too long and I ran out of patience" we would close the issue with "not planned" and an apology.

TL;DR Timestamps are cute, add to verbosity, and don't help all that much.

**Comment by xexyl** [comment]
I can't tell you how many times they have helped me!

But it's for those who develop of course. 

Nonetheless I would argue that they should only be done if there's the start and end message and I am unsure how useful those actually are.

**Comment by xexyl** [comment]
So the real question is should those messages be removed entirely. For some I think it might be useful but do we really need them for most rules?

**Comment by lcn2** [comment]
> So the real question is should those messages be removed entirely. For some I think it might be useful but do we really need them for most rules?

We do **NOT** think timestamps are needed for such "beginning" and "ending" messages.  They might have been fun during development, but now they only clutter and the enable people to whine about a performance of a command that they might have to run just once and produce data we cannot / will-not do anything about.  The time for "major development" in this repo has passed, and so any past use of "beginning" and "ending" messages having timestamps is in the past and is **NOT** needed now.

... yes, there will be, post IOCCC28, and update to `chkentry(1)`, but that is is an enhancement to one tool and is not "major development", even though the coding will be a fun challenge for you.  :-)

As for the "beginning" and "ending" messages w/o timestamps:

Such messages are useful for the higher level `Makefile` when calling sub-directory `foo/Makefile` actions.  However at the leaf of a build tree, they are not strictly needed.

For example, a "beginning" and "ending" message for the `dbg/` and `dyn_array/` directories (and by extension those repos) are **NOT** strictly needed.  If upper level Makefiles in the build tree use "beginning" and "ending" messages w/o timestamps when calling them, the actions can be inferred.  The upper level `Makefile` that invokes them (i.e., that invokes an action in `dbg/Makefile` and `dyn_array/Makefile`) **DOES** and **SHOULD** have a "beginning" and "ending" messages w/o timestamps.

A similar case of a "beginning" and "ending" message being **NOT** strictly needed for `soup/Makefile` and `test_ioccc/Makefile`.  Those are leaf nodes, so let the upper level Makefiles be the ones to issue "beginning" and "ending" messages w/o timestamps.

Now `jparse/Makefile` (and this the top level makefile for jparse repo), *SHOULD** have a "beginning" and "ending" messages w/o timestamps: especially as it needs to call `jparse/test_jparse/Makefile`) because **NOT** a leaf of the build tree.

We hope this clarifies.

**Comment by xexyl** [comment]
> > So the real question is should those messages be removed entirely. For some I think it might be useful but do we really need them for most rules?
> 
> 
> 
> We do **NOT** think timestamps are needed for such "beginning" and "ending" messages.  They might have been fun during development, but now they only clutter and the enable people to whine about a performance of a command that they might have to run just once and produce data we cannot / will-not do anything about.  The time for "major development" in this repo has passed, and so any past use of "beginning" and "ending" messages having timestamps is in the past and is **NOT** needed now.
> 
> 
> 
> ... yes, there will be, post IOCCC28, and update to `chkentry(1)`, but that is is an enhancement to one tool and is not "major development", even though the coding will be a fun challenge for you.  :-)
> 
> 
> 
> As for the "beginning" and "ending" messages w/o timestamps:
> 
> 
> 
> Such messages are useful for the higher level `Makefile` when calling sub-directory `foo/Makefile` actions.  However at the leaf of a build tree, they are not strictly needed.
> 
> 
> 
> For example, a "beginning" and "ending" message for the `dbg/` and `dyn_array/` directories (and by extension those repos) are **NOT** strictly needed.  If upper level Makefiles in the build tree use "beginning" and "ending" messages w/o timestamps when calling them, the actions can be inferred.  The upper level `Makefile` that invokes them (i.e., that invokes an action in `dbg/Makefile` and `dyn_array/Makefile`) **DOES** and **SHOULD** have a "beginning" and "ending" messages w/o timestamps.
> 
> 
> 
> A similar case of a "beginning" and "ending" message being **NOT** strictly needed for `soup/Makefile` and `test_ioccc/Makefile`.  Those are leaf nodes, so let the upper level Makefiles be the ones to issue "beginning" and "ending" messages w/o timestamps.
> 
> 
> 
> Now `jparse/Makefile` (and this the top level makefile for jparse repo), *SHOULD** have a "beginning" and "ending" messages w/o timestamps: especially as it needs to call `jparse/test_jparse/Makefile`) because **NOT** a leaf of the build tree.
> 
> 
> 
> We hope this clarifies, @xexyl.

Yes the problem is the subdirectories. 

I am afraid that I have to leave now but I would be happy to adjust things tomorrow. Easy to do. 

The question is do we want to have the start and end messages for subdirectories? But that introduces the problem of other repos. 

Please advise how you would like me to address that part and I will take care of it in all repos tomorrow. 

If you need you can hold off on the other repos PRs even if you did think at first they should have them. 

Back tomorrow!

**Comment by lcn2** [comment]
> The question is do we want to have the start and end messages for subdirectories? But that introduces the problem of other repos.

We do not see the problem.   Leaf makefiles do not need such messages.  The dbg and dyn_array Makefiles are leaf makefiles form the perspective of this repo.   They are also leaf makfiles for their own repo.  No problem here.

The `jparse/Makefile` is NOT a leaf makefile: not for this repo, nor for the jparse repo, so it can make good use of such messages.

Again, no problem here for both this repo and the jparse repo.

**Comment by xexyl** [comment]
Fixed the timestamp problem in 5e3bb2d6725a0c4646895bb1136d8e7a914c2be4. Have to look at the other stuff.

Now as far as the beginning / ending stuff in 'leaf' Makefiles: I did not add those but you did. Should they be removed? But then in that case it brings up the problem again with `jparse` (which incidentally the timestamps were removed from that too). I believe you brought up a possible solution but I have to locate it first.

**Comment by xexyl** [comment]
> > The question is do we want to have the start and end messages for subdirectories? But that introduces the problem of other repos.
> 
> We do not see the problem. Leaf makefiles do not need such messages. The dbg and dyn_array Makefiles are leaf makefiles form the perspective of this repo. They are also leaf makfiles for their own repo. No problem here.
> 
> The `jparse/Makefile` is NOT a leaf makefile: not for this repo, nor for the jparse repo, so it can make good use of such messages.
> 
> Again, no problem here for both this repo and the jparse repo.
> 

I'm not sure if I followed you right though. Do you mean that the `dbg` and `dyn_array` Makefiles should or should NOT have them? Since they're an outside repo I do not know.

Perhaps you just mean that `soup` and `test_ioccc` should have them removed? Please either advise or else take care of the messages (removing if necessary).


> > If you need you can hold off on the other repos PRs even if you did think at first they should have them.
> 
> We do not have any edits planned for the dbg repo and the dyn_array repo, nor for this repo.
> 

That's good to know.

> And as long as you don't introduce "beginning" and "ending" messages into the leaf node Makefiles, all will be fine.
> 

As noted I have removed the timestamps (even if they can be useful at times I can see it being overmuch esp in the subdirectories and even worse in subdirectories of the first subdirectories ... but then I also see that for the messages in general).


> And as long as a "virgin build" still works for this repo (i.e., **NOT** requiring any other repo to be installed), all will be fine.

I can't imagine it won't but I'll check again.

**Comment by lcn2** [comment]
>> We do not see the problem. Leaf makefiles do not need such messages. The dbg and dyn_array Makefiles are leaf makefiles form the perspective of this repo. They are also leaf makfiles for their own repo. No problem here.
>>
>> The jparse/Makefile is NOT a leaf makefile: not for this repo, nor for the jparse repo, so it can make good use of such messages.
>>
>> Again, no problem here for both this repo and the jparse repo.
>>
> I'm not sure if I followed you right though. Do you mean that the dbg and dyn_array Makefiles should or should NOT have them? 

Yes, @xexyl 

> Since they're an outside repo I do not know.

Those repos are outside of this repo, yes, but their contents are "copied" into this repo as sub-directories and thus become "inside" this repo. 

> Perhaps you just mean that soup and test_ioccc should have them removed? Please either advise or else take care of the messages (removing if necessary).

The `soup/Makefile` and `test_ioccc/Makefile` are leaf Makefiles, don't need them, and so should have them removed.

**Comment by xexyl** [comment]
> > > We do not see the problem. Leaf makefiles do not need such messages. The dbg and dyn_array Makefiles are leaf makefiles form the perspective of this repo. They are also leaf makfiles for their own repo. No problem here.
> > > The jparse/Makefile is NOT a leaf makefile: not for this repo, nor for the jparse repo, so it can make good use of such messages.
> > > Again, no problem here for both this repo and the jparse repo.
> > 
> > 
> > I'm not sure if I followed you right though. Do you mean that the dbg and dyn_array Makefiles should or should NOT have them?
> 
> Yes, @xexyl
> 
> > Since they're an outside repo I do not know.
> 
> Those repos are outside of this repo, yes, but their contents are "copied" into this repo as sub-directories and thus become "inside" this repo.
> 
> > Perhaps you just mean that soup and test_ioccc should have them removed? Please either advise or else take care of the messages (removing if necessary).
> 
> The `soup/Makefile` and `test_ioccc/Makefile` are leaf Makefiles, don't need them, and so should have them removed.

That's what I thought. Want me to do that or do you want to so I can take care of the other issue ?

**Comment by lcn2** [comment]
> > > > We do not see the problem. Leaf makefiles do not need such messages. The dbg and dyn_array Makefiles are leaf makefiles form the perspective of this repo. They are also leaf makfiles for their own repo. No problem here.
> > > > The jparse/Makefile is NOT a leaf makefile: not for this repo, nor for the jparse repo, so it can make good use of such messages.
> > > > Again, no problem here for both this repo and the jparse repo.
> > > 
> > > 
> > > I'm not sure if I followed you right though. Do you mean that the dbg and dyn_array Makefiles should or should NOT have them?
> > 
> > 
> > Yes, @xexyl
> > > Since they're an outside repo I do not know.
> > 
> > 
> > Those repos are outside of this repo, yes, but their contents are "copied" into this repo as sub-directories and thus become "inside" this repo.
> > > Perhaps you just mean that soup and test_ioccc should have them removed? Please either advise or else take care of the messages (removing if necessary).
> > 
> > 
> > The `soup/Makefile` and `test_ioccc/Makefile` are leaf Makefiles, don't need them, and so should have them removed.
> 
> That's what I thought. Want me to do that or do you want to so I can take care of the other issue ?

We could but as you are dealing with [GH-issuecomment-2336454677](https://github.com/ioccc-src/mkiocccentry/issues/723#issuecomment-2336454677) it would be better to not complicate things by editing in this repo.  So please take care of this as you address the build issue.
